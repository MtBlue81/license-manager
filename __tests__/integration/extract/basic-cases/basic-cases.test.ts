import { ExtractArgs, extract } from "@/extract";
import { consoleMock } from "__tests__/helpers/consoleMock";
import { processMock } from "__tests__/helpers/processMock";
import { fs as memfs, vol } from "memfs";
import path from "path";
import pc from "picocolors";
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";

vi.mock("fs");

describe("extract : basic-cases", () => {
  beforeEach(() => {
    processMock(__dirname);
    consoleMock();
  });

  afterEach(() => {
    vol.reset();
  });

  const extractDefaultOption: ExtractArgs = {
    workspace: "",
    cwd: __dirname,
    query: "",
    extractLicenses: [],
    excludePackages: [],
    output: "",
    json: false,
  };

  it("extract all licenses", async () => {
    await extract({ ...extractDefaultOption });

    const expectedOutputPath = path.resolve(__dirname, "licenses.txt");

    expect(console.log).toBeCalledWith(pc.green(`âœ… Extracted to ${expectedOutputPath}`));

    expect(memfs.readFileSync(expectedOutputPath).toString("utf-8")).toMatchSnapshot();
  });

  it("extract specified licenses", async () => {
    await extract({
      ...extractDefaultOption,
      extractLicenses: ["MIT", "Apache-2.0"],
    });

    const expectedOutputPath = path.resolve(__dirname, "licenses.txt");

    expect(console.log).toBeCalledWith(pc.green(`âœ… Extracted to ${expectedOutputPath}`));

    expect(memfs.readFileSync(expectedOutputPath).toString("utf-8")).toMatchSnapshot();
  });

  it("extract licenses by excluding specified packages", async () => {
    await extract({
      ...extractDefaultOption,
      excludePackages: ["baz"],
    });

    const expectedOutputPath = path.resolve(__dirname, "licenses.txt");

    expect(console.log).toBeCalledWith(pc.green(`âœ… Extracted to ${expectedOutputPath}`));

    expect(memfs.readFileSync(expectedOutputPath).toString("utf-8")).toMatchSnapshot();
  });

  it("license text not found in some packages", async () => {
    await expect(() =>
      extract({
        ...extractDefaultOption,
        query: ":root > [name^=@invalid/]",
      })
    ).rejects.toThrowError("process.exit()");

    const expectedOutputPath = path.resolve(__dirname, "licenses.txt");

    expect(console.error).toBeCalledWith(pc.red("ðŸš¨ License text not found"));
    expect(console.error).toBeCalledWith(pc.red("@invalid/baz@1.2.3"));
    expect(console.error).toBeCalledWith(pc.red("@invalid/foo@1.2.3"));

    expect(memfs.existsSync(expectedOutputPath)).toBe(false);
  });

  it("can specify query", async () => {
    await extract({
      ...extractDefaultOption,
      query: ":root > [name^=@dev/]",
    });

    const expectedOutputPath = path.resolve(__dirname, "licenses.txt");

    expect(console.log).toBeCalledWith(pc.green(`âœ… Extracted to ${expectedOutputPath}`));

    expect(memfs.readFileSync(expectedOutputPath).toString("utf-8")).toMatchSnapshot();
  });

  it("output in json format", async () => {
    await extract({
      ...extractDefaultOption,
      json: true,
    });

    const expectedOutputPath = path.resolve(__dirname, "licenses.json");

    expect(console.log).toBeCalledWith(pc.green(`âœ… Extracted to ${expectedOutputPath}`));

    const json = JSON.parse(memfs.readFileSync(expectedOutputPath).toString("utf-8"));

    const matchPattern = expect.objectContaining({
      name: expect.any(String),
      version: expect.any(String),
      license: expect.any(String),
      path: expect.any(String),
      realpath: expect.any(String),
      licenseTextPath: expect.any(String),
      licenseText: expect.any(String),
    });

    expect(json[0]).toMatchObject(matchPattern);
    expect(json[1]).toMatchObject(matchPattern);
    expect(json[2]).toMatchObject(matchPattern);
  });
});
