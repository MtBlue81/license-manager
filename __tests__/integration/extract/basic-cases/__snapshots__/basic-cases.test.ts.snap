// Vitest Snapshot v1

exports[`extract : basic-cases > can specify query 1`] = `
"@dev/bar
license: BSD
@dev/bar - license text - line1
@dev/bar - license text - line2



@dev/baz
license: Apache-2.0
@dev/baz - license text - line1
@dev/baz - license text - line2



@dev/foo
license: MIT
@dev/foo - license text - line1
@dev/foo - license text - line2
"
`;

exports[`extract : basic-cases > extract all licenses 1`] = `
"bar
license: BSD
bar - license text - line1
bar - license text - line2



baz
license: Apache-2.0
baz - license text - line1
baz - license text - line2



foo
license: MIT
foo - license text - line1
foo - license text - line2



qux
license: MIT
qux - license text - line1
qux - license text - line2
"
`;

exports[`extract : basic-cases > extract licenses by excluding specified packages 1`] = `
"bar
license: BSD
bar - license text - line1
bar - license text - line2



foo
license: MIT
foo - license text - line1
foo - license text - line2



qux
license: MIT
qux - license text - line1
qux - license text - line2
"
`;

exports[`extract : basic-cases > extract specified licenses 1`] = `
"baz
license: Apache-2.0
baz - license text - line1
baz - license text - line2



foo
license: MIT
foo - license text - line1
foo - license text - line2



qux
license: MIT
qux - license text - line1
qux - license text - line2
"
`;
